README.md
# qbcall

nexmo-python/CHANGES.md
# 2.4.0
* Application V2 API added under `Client.application_v2`
* Existing application methods under `Client` are now deprecated.

# 2.3.0
* Explicit parameter list for the `nexmo.Client` constructor. **This may cause errors in code passing incorrect or spurious arguments to the Client constructor.**
* Secret Management
* Support for Authorization header authentication.

# 2.2.0
* Add support for `redact_transaction`.

# 2.1.0
* Add support for `get_recording`
* Add support for SMS conversion
* Add debug logging for most calls, under the 'nexmo' logger.
* Internal refactoring (affects only private methods.)

# 2.0.0
* Drop support for Python 3.3 (in line with the cryptography library we depend upon)
* Ensure timestamp is added the params list if signing requests
* Avoid value injection in signature auth.
* Add support for different hashes for signature generation (thanks @trancee!)
* Tests ported to pytest

# 1.5.0

* Add ability to provide a file path as private_key param no the nexmo.Client constructor

* Add send/stop endpoints for audio/speech/dtmf

* Add new number insight endpoints

# 1.4.0

* Add new Voice API call methods

* Add Application API methods

* Add check_signature method for checking callback signatures

* Deprecate old Verify API methods

# 1.3.0

* Add get_sms_pricing method

* Add get_voice_pricing method

* Add get_event_alert_numbers method to get opt-in/opt-out numbers

* Add resubscribe_event_alert_number method to opt-in a number

* Add more clearly named methods for Verify API

* Add app_name and app_version options

# 1.2.0

* Add topup method

* Add update_settings method

* Add api_host attribute

* Add ClientError and ServerError classes

# 1.1.0

* Move repository to https://github.com/Nexmo/nexmo-python

* Add get_basic_number_insight method for Number Insight Basic API

* Add get_number_insight method for Number Insight Standard API

* Add User-Agent header to requests

# 1.0.3

* Change license from LGPL-3.0 to MIT

# 1.0.2

* Remove merge helper function

# 1.0.1

* Python 3.x fixes

# 1.0.0

* First version!


nexmo-python/tests/util.py
import os.path
import re

import pytest


try:
    from urllib.parse import urlparse, parse_qs
except ImportError:
    from urlparse import urlparse, parse_qs

import responses


def request_body():
    return responses.calls[0].request.body


def request_query():
    return urlparse(responses.calls[0].request.url).query


def request_params():
    """ Obtain the query params, as a dict. """
    return parse_qs(request_query())


def request_headers():
    return responses.calls[0].request.headers


def request_user_agent():
    return responses.calls[0].request.headers["User-Agent"]


def request_authorization():
    return responses.calls[0].request.headers["Authorization"].decode("utf-8")


def request_content_type():
    return responses.calls[0].request.headers["Content-Type"]


def stub(method, url, fixture_path=None, status_code=200):
    body = load_fixture(fixture_path) if fixture_path else '{"key":"value"}'
    responses.add(
        method, url, body=body, status=status_code, content_type="application/json"
    )


def stub_bytes(method, url):
    responses.add(method, url, body=b"THISISANMP3", status=200)


def assert_re(pattern, string):
    __tracebackhide__ = True
    if not re.search(pattern, string):
        pytest.fail("Cannot find pattern {!r} in {!r}".format(pattern, string))


def assert_basic_auth():
    params = request_params()
    assert "api_key" not in params
    assert "api_secret" not in params
    assert (
        request_headers()["Authorization"]
        == "Basic bmV4bW8tYXBpLWtleTpuZXhtby1hcGktc2VjcmV0"
    )


def load_fixture(fixture_path):
    return open(os.path.join(os.path.dirname(__file__), "data", fixture_path)).read()


nexmo-python/tests/conftest.py
import os
import os.path
import platform

import pytest


# Ensure our client isn't being configured with real values!
os.environ.clear()


def read_file(path):
    with open(os.path.join(os.path.dirname(__file__), path)) as input_file:
        return input_file.read()


class DummyData(object):
    def __init__(self):
        import nexmo

        self.api_key = "nexmo-api-key"
        self.api_secret = "nexmo-api-secret"
        self.signature_secret = "secret"
        self.application_id = "nexmo-application-id"
        self.private_key = read_file("data/private_key.txt")
        self.public_key = read_file("data/public_key.txt")
        self.user_agent = "nexmo-python/{} python/{}".format(
            nexmo.__version__, platform.python_version()
        )


@pytest.fixture(scope="session")
def dummy_data():
    return DummyData()


@pytest.fixture
def client(dummy_data):
    import nexmo

    return nexmo.Client(
        key=dummy_data.api_key,
        secret=dummy_data.api_secret,
        application_id=dummy_data.application_id,
        private_key=dummy_data.private_key,
    )


nexmo-python/tests/test_sms.py
import nexmo
from util import *


@responses.activate
def test_send_message(client, dummy_data):
    stub(responses.POST, "https://rest.nexmo.com/sms/json")

    params = {"from": "Python", "to": "447525856424", "text": "Hey!"}

    assert isinstance(client.send_message(params), dict)
    assert request_user_agent() == dummy_data.user_agent
    assert "from=Python" in request_body()
    assert "to=447525856424" in request_body()
    assert "text=Hey%21" in request_body()


@responses.activate
def test_authentication_error(client):
    responses.add(responses.POST, "https://rest.nexmo.com/sms/json", status=401)

    with pytest.raises(nexmo.AuthenticationError):
        client.send_message({})


@responses.activate
def test_client_error(client):
    responses.add(responses.POST, "https://rest.nexmo.com/sms/json", status=400)

    with pytest.raises(nexmo.ClientError) as excinfo:
        client.send_message({})
    excinfo.match(r"400 response from rest.nexmo.com")


@responses.activate
def test_server_error(client):
    responses.add(responses.POST, "https://rest.nexmo.com/sms/json", status=500)

    with pytest.raises(nexmo.ServerError) as excinfo:
        client.send_message({})
    excinfo.match(r"500 response from rest.nexmo.com")


@responses.activate
def test_submit_sms_conversion(client):
    responses.add(
        responses.POST, "https://api.nexmo.com/conversions/sms", status=200, body=b"OK"
    )

    client.submit_sms_conversion("a-message-id")
    assert "message-id=a-message-id" in request_body()
    assert "timestamp" in request_body()


nexmo-python/src/nexmo/_internal.py
import logging

from requests.sessions import Session

from .errors import AuthenticationError, ClientError, ServerError

try:
    from json import JSONDecodeError
except ImportError:
    JSONDecodeError = ValueError

logger = logging.getLogger("nexmo")


class BasicAuthenticatedServer(object):
    def __init__(self, host, user_agent, api_key, api_secret):
        self._host = host
        self._session = session = Session()
        session.auth = (api_key, api_secret)  # Basic authentication.
        session.headers.update({"User-Agent": user_agent})

    def _uri(self, path):
        return "{host}{path}".format(host=self._host, path=path)

    def get(self, path, params=None, headers=None):
        return self._parse(
            self._session.get(self._uri(path), params=params, headers=headers)
        )

    def post(self, path, body=None, headers=None):
        return self._parse(
            self._session.post(self._uri(path), json=body, headers=headers)
        )

    def put(self, path, body=None, headers=None):
        return self._parse(
            self._session.put(self._uri(path), json=body, headers=headers)
        )

    def delete(self, path, body=None, headers=None):
        return self._parse(
            self._session.delete(self._uri(path), json=body, headers=headers)
        )

    def _parse(self, response):
        logger.debug("Response headers %r", response.headers)
        if response.status_code == 401:
            raise AuthenticationError()
        elif response.status_code == 204:
            return None
        elif 200 <= response.status_code < 300:
            return response.json()
        elif 400 <= response.status_code < 500:
            logger.warning(
                "Client error: %s %r", response.status_code, response.content
            )
            message = "{code} response".format(code=response.status_code)
            # Test for standard error format:
            try:
                error_data = response.json()
                if (
                    "type" in error_data
                    and "title" in error_data
                    and "detail" in error_data
                ):
                    message = "{title}: {detail} ({type})".format(
                        title=error_data["title"],
                        detail=error_data["detail"],
                        type=error_data["type"],
                    )
            except JSONDecodeError:
                pass
            raise ClientError(message)
        elif 500 <= response.status_code < 600:
            logger.warning(
                "Server error: %s %r", response.status_code, response.content
            )
            message = "{code} response".format(code=response.status_code)
            raise ServerError(message)


class ApplicationV2(object):
    """
    Provides Application API v2 functionality.

    Don't instantiate this class yourself, access it via :py:attr:`nexmo.Client.application_v2`
    """

    def __init__(self, api_server):
        self._api_server = api_server

    def create_application(self, application_data):
        """
        Create an application using the provided `application_data`.

        :param dict application_data: A JSON-style dict describing the application to be created.

        >>> client.application_v2.create_application({ 'name': 'My Cool App!' })

        Details of the `application_data` dict are described at https://developer.nexmo.com/api/application.v2#createApplication
        """
        return self._api_server.post("/v2/applications", application_data)

    def get_application(self, application_id):
        """
        Get application details for the application with `application_id`.

        The format of the returned dict is described at https://developer.nexmo.com/api/application.v2#getApplication

        :param str application_id: The application ID.
        :rtype: dict
        """

        return self._api_server.get(
            "/v2/applications/{application_id}".format(application_id=application_id),
            headers={"content-type": "application/json"},
        )

    def update_application(self, application_id, params):
        """
        Update the application with `application_id` using the values provided in `params`.


        """
        return self._api_server.put(
            "/v2/applications/{application_id}".format(application_id=application_id),
            params,
        )

    def delete_application(self, application_id):
        """
        Delete the application with `application_id`.
        """

        self._api_server.delete(
            "/v2/applications/{application_id}".format(application_id=application_id),
            headers={"content-type": "application/json"},
        )

    def list_applications(self, page_size=None, page=None):
        """
        List all applications for your account.

        Results are paged, so each page will need to be requested to see all applications.

        :param int page_size: The number of items in the page to be returned
        :param int page: The page number of the page to be returned.
        """
        params = _filter_none_values({"page_size": page_size, "page": page})

        return self._api_server.get(
            "/v2/applications",
            params=params,
            headers={"content-type": "application/json"},
        )


def _filter_none_values(d):
    return {k: v for k, v in d.items() if v is not None}


def _format_date_param(params, key, format="%Y-%m-%d %H:%M:%S"):
    """
    Utility function to convert datetime values to strings.

    If the value is already a str, or is not in the dict, no change is made.

    :param params: A `dict` of params that may contain a `datetime` value.
    :param key: The datetime value to be converted to a `str`
    :param format: The `strftime` format to be used to format the date. The default value is '%Y-%m-%d %H:%M:%S'
    """
    if key in params:
        param = params[key]
        if hasattr(param, "strftime"):
            params[key] = param.strftime(format)
